src/reverse.jl

```diff
@@ -5,105 +5,126 @@
 # pos [i1, i2, i3, ... , d{x} - i{x} + 1, ..., i{n}] where d{x} is the size of dimension x
 
 # out-of-place version, copying a single value per thread from input to output
-function _reverse(input::AnyCuArray{T, N}, output::AnyCuArray{T, N};
-                  dims::Integer=1) where {T, N}
+function _reverse(input::AnyCuArray{T, N}, output::AnyCuArray{T, N}; dims=1:ndims(input)) where {T, N}
     @assert size(input) == size(output)
-    shape = [size(input)...]
-    numelemsinprevdims = prod(shape[1:dims-1])
-    numelemsincurrdim = shape[dims]
+    rev_dims = ntuple((d)-> d in dims && size(input, d) > 1, N)
+    ref = size(input) .+ 1
+    # converts an ND-index in the data array to the linear index
+    lin_idx = LinearIndices(input)
+    # converts a linear index in a reduced array to an ND-index, but using the reduced size
+    nd_idx = CartesianIndices(input)
 
     function kernel(input::AbstractArray{T, N}, output::AbstractArray{T, N}) where {T, N}
         offset_in = blockDim().x * (blockIdx().x - 1i32)
-
         index_in = offset_in + threadIdx().x
 
         if index_in <= length(input)
-            element = @inbounds input[index_in]
-
-            # the index of an element in the original array along dimension that we will flip
-            #assume(numelemsinprevdims > 0)
-            #assume(numelemsincurrdim > 0)
-            ik = ((cld(index_in, numelemsinprevdims) - 1) % numelemsincurrdim) + 1
-
-            index_out = index_in + (numelemsincurrdim - 2ik + 1) * numelemsinprevdims
-
-            @inbounds output[index_out] = element
+            idx = Tuple(nd_idx[index_in])
```

Comment:
You need to extend the `@inbounds` to cover the `nd_idx` and `lin_idx` indexing too.

Reply:
Thanx! Great idea.